File: apps/habiter/android/app/src/test/kotlin/com/habiter/app/RuntimeFeatureStateTest.kt
package com.habiter.app

import com.habiter.app.runtime.RuntimeFeatureState
import org.junit.Assert.assertEquals
import org.junit.Test

class RuntimeFeatureStateTest {
    @Test
    fun `runtime only stops when both features are disabled`() {
        val cases = listOf(
            Triple(false, false, false),
            Triple(true, false, true),
            Triple(false, true, true),
            Triple(true, true, true),
        )

        cases.forEach { (reminders, appBlock, expected) ->
            assertEquals(
                expected,
                RuntimeFeatureState(reminders, appBlock).shouldRun,
            )
        }
    }

    @Test
    fun `notification text describes active features without App Lock ownership`() {
        assertEquals(
            "Reminders are running locally in the background.",
            RuntimeFeatureState(remindersEnabled = true, appBlockEnabled = false)
                .notificationText(),
        )
        assertEquals(
            "Reminders and App Block are active.",
            RuntimeFeatureState(remindersEnabled = true, appBlockEnabled = true)
                .notificationText(),
        )
    }
}

File: apps/habiter/test/features/runtime/runtime_feature_state_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:habiter/features/runtime/domain/runtime_feature_state.dart';

void main() {
  test(
    'runtime feature-state matrix only stops when both features are off',
    () {
      const cases = <(bool, bool, bool)>[
        (false, false, false),
        (true, false, true),
        (false, true, true),
        (true, true, true),
      ];

      for (final (reminders, appBlock, shouldRun) in cases) {
        final state = RuntimeFeatureState(
          remindersEnabled: reminders,
          appBlockEnabled: appBlock,
        );
        expect(
          state.shouldRun,
          shouldRun,
          reason: 'reminders=$reminders appBlock=$appBlock',
        );
      }
    },
  );

  test('active features and serialized shape remain explicit', () {
    const state = RuntimeFeatureState(
      remindersEnabled: true,
      appBlockEnabled: false,
    );

    expect(state.activeFeatures, <RuntimeFeature>{RuntimeFeature.reminders});
    expect(RuntimeFeatureState.fromMap(state.toMap()).toMap(), state.toMap());
  });
}
